File: android/src/main/kotlin/cloud/trustpin/flutter/sdk/MethodHandlers.kt
package cloud.trustpin.flutter.sdk

import cloud.trustpin.kotlin.sdk.TrustPin
import cloud.trustpin.kotlin.sdk.TrustPinConfiguration
import cloud.trustpin.kotlin.sdk.TrustPinError
import cloud.trustpin.kotlin.sdk.fromAssets
import cloud.trustpin.kotlin.sdk.withAndroidStorage
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel.Result

internal fun TrustPinSDKPlugin.handleSetup(call: MethodCall, result: Result) {
    execute(coroutineScope, result, ErrorCode.SETUP) {
        val organizationId = call.requireString(Arg.ORGANIZATION_ID)
        val projectId = call.requireString(Arg.PROJECT_ID)
        val publicKey = call.requireString(Arg.PUBLIC_KEY)
        val instanceId = call.optionalString(Arg.INSTANCE_ID)
        val configurationURL = call.optionalURL(Arg.CONFIGURATION_URL)
        val mode = call.optionalString(Arg.MODE).toTrustPinMode()
        // Required by the native SDK 5.0.0+ `withAndroidStorage` hardening
        // chain. Absent only if the plugin was used before onAttachedToEngine
        // ran, which indicates a Flutter engine attachment bug.
        val context = applicationContext ?: throw TrustPinError.InvalidProjectConfig

        suspend {
            val configuration = TrustPinConfiguration(
                organizationId = organizationId,
                projectId = projectId,
                publicKey = publicKey,
                mode = mode,
                configurationURL = configurationURL
            ).withAndroidStorage(context)
            trustPinInstance(instanceId).setup(configuration)
            null
        }
    }
}

internal fun TrustPinSDKPlugin.handleSetupWithNativeBundle(call: MethodCall, result: Result) {
    execute(coroutineScope, result, ErrorCode.SETUP) {
        val fileName = call.optionalString(Arg.ANDROID_FILE_NAME)
        val instanceId = call.optionalString(Arg.INSTANCE_ID)
        val context = applicationContext ?: throw TrustPinError.InvalidProjectConfig

        suspend {
            // `fromAssets` chains `withAndroidStorage(context)` internally, so
            // no extra hardening step is required for this path. When the
            // caller passes no filename we let the native SDK pick its own
            // default (`trustpin.json`) instead of duplicating the constant.
            val configuration = if (fileName == null) {
                TrustPinConfiguration.fromAssets(context)
            } else {
                TrustPinConfiguration.fromAssets(context, fileName)
            }
            trustPinInstance(instanceId).setup(configuration)
            null
        }
    }
}

internal fun TrustPinSDKPlugin.handleVerify(call: MethodCall, result: Result) {
    execute(coroutineScope, result, ErrorCode.VERIFY) {
        val domain = call.requireString(Arg.DOMAIN)
        val pem = call.requireString(Arg.CERTIFICATE)
        val instanceId = call.optionalString(Arg.INSTANCE_ID)
        val certificate = parsePemCertificate(pem)

        suspend {
            trustPinInstance(instanceId).verify(domain, certificate)
            null
        }
    }
}

internal fun TrustPinSDKPlugin.handleSetLogLevel(call: MethodCall, result: Result) {
    execute(coroutineScope, result, ErrorCode.SET_LOG_LEVEL) {
        val logLevel = call.requireString(Arg.LOG_LEVEL).toTrustPinLogLevel()
        val instanceId = call.optionalString(Arg.INSTANCE_ID)

        suspend {
            trustPinInstance(instanceId).setLogLevel(logLevel)
            null
        }
    }
}

internal fun TrustPinSDKPlugin.handleFetchCertificate(call: MethodCall, result: Result) {
    execute(coroutineScope, result, ErrorCode.FETCH_CERTIFICATE) {
        val host = call.requireString(Arg.HOST)
        val port = call.optionalInt(Arg.PORT) ?: DEFAULT_TLS_PORT
        val timeoutMs = call.optionalInt(Arg.TIMEOUT_MS)
        val instanceId = call.optionalString(Arg.INSTANCE_ID)

        suspend {
            val trustPin = trustPinInstance(instanceId)
            // A null/non-positive timeout defers to the native SDK's default;
            // the SDK raises TrustPinError.Timeout when the bound is exceeded.
            if (timeoutMs != null && timeoutMs > 0) {
                trustPin.fetchCertificate(host, port, timeoutMs.toLong())
            } else {
                trustPin.fetchCertificate(host, port)
            }
        }
    }
}

internal fun TrustPinSDKPlugin.handleValidateConnection(call: MethodCall, result: Result) {
    execute(coroutineScope, result, ErrorCode.VALIDATE_CONNECTION) {
        val host = call.requireString(Arg.HOST)
        val port = call.optionalInt(Arg.PORT) ?: DEFAULT_TLS_PORT
        val timeoutMs = call.optionalInt(Arg.TIMEOUT_MS)
        val instanceId = call.optionalString(Arg.INSTANCE_ID)

        suspend {
            val trustPin = trustPinInstance(instanceId)
            if (timeoutMs != null && timeoutMs > 0) {
                // The caller's timeout bounds the composed operation. The
                // native API takes one bound per call, so the verify phase
                // gets whatever the fetch phase left of the budget.
                val start = System.nanoTime()
                val pem = trustPin.fetchCertificate(host, port, timeoutMs.toLong())
                val elapsedMs = (System.nanoTime() - start) / 1_000_000
                val remainingMs = timeoutMs - elapsedMs
                if (remainingMs <= 0) throw TrustPinError.Timeout
                trustPin.verify(host, parsePemCertificate(pem), remainingMs)
            } else {
                val pem = trustPin.fetchCertificate(host, port)
                trustPin.verify(host, parsePemCertificate(pem))
            }
            null
        }
    }
}

internal fun TrustPinSDKPlugin.handleAwaitConfiguration(call: MethodCall, result: Result) {
    execute(coroutineScope, result, ErrorCode.AWAIT_CONFIGURATION) {
        val timeoutMs = call.optionalInt(Arg.TIMEOUT_MS)
        val instanceId = call.optionalString(Arg.INSTANCE_ID)

        suspend {
            val trustPin = trustPinInstance(instanceId)
            // The native API takes a millisecond deadline. A null/non-positive
            // value defers to the SDK's default timeout.
            if (timeoutMs != null && timeoutMs > 0) {
                trustPin.awaitConfiguration(timeoutMs.toLong())
            } else {
                trustPin.awaitConfiguration()
            }
            null
        }
    }
}

internal fun TrustPinSDKPlugin.handleIsConfigurationLoaded(call: MethodCall, result: Result) {
    execute(coroutineScope, result, ErrorCode.AWAIT_CONFIGURATION) {
        val instanceId = call.optionalString(Arg.INSTANCE_ID)

        suspend {
            trustPinInstance(instanceId).isConfigurationLoaded
        }
    }
}

/**
 * Returns the named TrustPin instance, or the default one when [id] is null
 * or empty.
 */
internal fun trustPinInstance(id: String?): TrustPin =
    if (id.isNullOrEmpty()) TrustPin.default else TrustPin.instance(id)

File: macos/trustpin_sdk/Sources/trustpin_sdk/MethodHandlers.swift
@preconcurrency import FlutterMacOS
import Foundation
import TrustPinKit

extension TrustPinSDKPlugin {

    func handleSetup(_ call: FlutterMethodCall, box: ResultBox) {
        execute(box, defaultCode: ErrorCode.setup) {
            let args = try Arguments(call)
            let organizationId = try args.requireString(Arg.organizationId)
            let projectId = try args.requireString(Arg.projectId)
            let publicKey = try args.requireString(Arg.publicKey)
            let instanceId = args.optionalString(Arg.instanceId)
            let configurationURL = try args.optionalURL(Arg.configurationURL)
            let mode = TrustPinMode(args.optionalString(Arg.mode))

            return {
                let configuration = TrustPinConfiguration(
                    organizationId: organizationId,
                    projectId: projectId,
                    publicKey: publicKey,
                    mode: mode,
                    configurationURL: configurationURL
                )
                let trustPin = try TrustPinSDKPlugin.instance(id: instanceId)
                try await trustPin.setup(configuration)
                return nil
            }
        }
    }

    func handleSetupWithNativeBundle(_ call: FlutterMethodCall, box: ResultBox) {
        execute(box, defaultCode: ErrorCode.setup) {
            let args = try Arguments(call)
            let fileName = args.optionalString(Arg.macosFileName)
            let instanceId = args.optionalString(Arg.instanceId)

            return {
                // When the caller passes no filename we let the native SDK
                // pick its own default (`TrustPin-Info.plist`) instead of
                // duplicating the constant on the Flutter side.
                let configuration: TrustPinConfiguration
                if let fileName {
                    configuration = try TrustPinConfiguration.fromPlist(.main, fileName: fileName)
                } else {
                    configuration = try TrustPinConfiguration.fromPlist()
                }
                let trustPin = try TrustPinSDKPlugin.instance(id: instanceId)
                try await trustPin.setup(configuration)
                return nil
            }
        }
    }

    func handleVerify(_ call: FlutterMethodCall, box: ResultBox) {
        execute(box, defaultCode: ErrorCode.verify) {
            let args = try Arguments(call)
            let domain = try args.requireString(Arg.domain)
            let certificate = try args.requireString(Arg.certificate)
            let instanceId = args.optionalString(Arg.instanceId)

            return {
                let trustPin = try TrustPinSDKPlugin.instance(id: instanceId)
                try await trustPin.verify(domain: domain, certificate: certificate)
                return nil
            }
        }
    }

    func handleSetLogLevel(_ call: FlutterMethodCall, box: ResultBox) {
        execute(box, defaultCode: ErrorCode.setLogLevel) {
            let args = try Arguments(call)
            let logLevel = TrustPinLogLevel(try args.requireString(Arg.logLevel))
            let instanceId = args.optionalString(Arg.instanceId)

            return {
                let trustPin = try TrustPinSDKPlugin.instance(id: instanceId)
                trustPin.set(logLevel: logLevel)
                return nil
            }
        }
    }

    func handleFetchCertificate(_ call: FlutterMethodCall, box: ResultBox) {
        execute(
            box,
            defaultCode: ErrorCode.fetchCertificate,
            timeoutMessage: "Timed out fetching certificate"
        ) {
            let args = try Arguments(call)
            let host = try args.requireString(Arg.host)
            let port = args.optionalInt(Arg.port) ?? defaultTLSPort
            let timeoutMs = args.optionalInt(Arg.timeoutMs)
            let instanceId = args.optionalString(Arg.instanceId)

            return {
                try await withOptionalTimeout(milliseconds: timeoutMs) { () async throws -> String in
                    let trustPin = try TrustPinSDKPlugin.instance(id: instanceId)
                    return try await trustPin.fetchCertificate(host: host, port: port)
                }
            }
        }
    }

    func handleValidateConnection(_ call: FlutterMethodCall, box: ResultBox) {
        execute(
            box,
            defaultCode: ErrorCode.validateConnection,
            timeoutMessage: "Timed out validating connection"
        ) {
            let args = try Arguments(call)
            let host = try args.requireString(Arg.host)
            let port = args.optionalInt(Arg.port) ?? defaultTLSPort
            let timeoutMs = args.optionalInt(Arg.timeoutMs)
            let instanceId = args.optionalString(Arg.instanceId)

            return {
                try await withOptionalTimeout(milliseconds: timeoutMs) {
                    let trustPin = try TrustPinSDKPlugin.instance(id: instanceId)
                    let pem = try await trustPin.fetchCertificate(host: host, port: port)
                    try await trustPin.verify(domain: host, certificate: pem)
                }
                return nil
            }
        }
    }

    func handleAwaitConfiguration(_ call: FlutterMethodCall, box: ResultBox) {
        execute(
            box,
            defaultCode: ErrorCode.awaitConfiguration,
            timeoutMessage: "Timed out awaiting configuration"
        ) {
            let args = try Arguments(call)
            let timeoutMs = args.optionalInt(Arg.timeoutMs)
            let instanceId = args.optionalString(Arg.instanceId)

            return {
                let trustPin = try TrustPinSDKPlugin.instance(id: instanceId)
                if let timeoutMs, timeoutMs > 0 {
                    try await trustPin.awaitConfiguration(
                        timeout: TimeInterval(timeoutMs) / 1000.0
                    )
                } else {
                    try await trustPin.awaitConfiguration()
                }
                return nil
            }
        }
    }

    func handleIsConfigurationLoaded(_ call: FlutterMethodCall, box: ResultBox) {
        execute(box, defaultCode: ErrorCode.awaitConfiguration) {
            let args = try Arguments(call)
            let instanceId = args.optionalString(Arg.instanceId)

            return {
                let trustPin = try TrustPinSDKPlugin.instance(id: instanceId)
                return await trustPin.isConfigurationLoaded
            }
        }
    }

    /// Returns the named TrustPin instance, or the default one when `id` is
    /// nil or empty. Kept as a static method so it never captures `self`
    /// from inside a `@Sendable` closure.
    static func instance(id: String?) throws -> TrustPin {
        guard let id, !id.isEmpty else { return TrustPin.default }
        return try TrustPin.instance(id: id)
    }
}
